src/Adapter/Order/CommandHandler/CancelOrderProductHandler.php

```diff
@@ -0,0 +1,173 @@
+<?php
+/**
+ * 2007-2019 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Order\CommandHandler;
+
+use Cart;
+use Configuration;
+use Context;
+use Customization;
+use Hook;
+use Order;
+use OrderCarrier;
+use OrderDetail;
+use PrestaShop\PrestaShop\Core\Domain\Order\Command\CancelOrderProductCommand;
+use PrestaShop\PrestaShop\Core\Domain\Order\CommandHandler\CancelOrderProductHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\EmptyProductSelectionException;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\InvalidCancelQuantityException;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\OrderException;
+use StockAvailable;
+use Validate;
+
+/**
+ * @internal
+ */
+final class CancelOrderProductHandler extends AbstractOrderCommandHandler implements CancelOrderProductHandlerInterface
+{
+    private $translator;
+
+    /**
+     * Legacy code for product cancellation handling in order page
+     */
+    public function handle(CancelOrderProductCommand $command)
```

Comment:
I think the whole handler code can be simplified through a few modifs:
- the invalid exceptions (negative quantity) should be managed in the command/valueobjects rather than in the handler
- you can use the the `OrderRefundCalculator` service to manage most calculations and parameters interpretation
- I'm not sure it's useful to separate the regular products and the customization in two separate arrays: you can simply loop through the OrderDetail list and check if `$orderDetail->id_customization` is present (see https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-61dbf27f8c77b84938556c403480bfe9R177 and https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-07c00c621d3ca10645633744ed48d14dR74)